### ISSUE TITLE
ENH: Implementation of a `scipy.fft.next_fast_len`-equivalent for `numpy.fft`

### ISSUE THREAD

**Issue by MothNik** [numpy/numpy#27987]
### Proposed new feature or change:

Scipy's `fft` package offers the function [`scipy.fft.next_fast_len`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.next_fast_len.html) to find the next fast length to which the input vector needs to be padded for a fast Discrete Fourier Transform.

It would be great to have such a feature for NumPy since NumPy's `fft` package is significantly faster than SciPy's and a performance penalty coming from an unfortunate input vector size thus hurts a lot.

**Comment by MothNik** [comment]
Thanks @mattip for the question!
It seems that SciPy's wrapping introduces a lot more overhead and for the small input sizes I'm working with (around 1000) this is significant.

<img width="521" alt="image" src="https://github.com/user-attachments/assets/bd121874-ccd3-4ba3-862e-4cbb008e9356" />

The plot was created on MacOS with `perfplot`.

```python
import numpy as np
import perfplot
from scipy.fft import fft as scipy_fft

perfplot.show(
    setup=lambda n: np.random.rand(n),
    n_range=[2**k for k in range(20)],
    kernels=[
        lambda a: np.fft.fft(a),
        lambda a: scipy_fft(a),
    ],
    labels=["np.fft.fft", "scipy.fft.fft"],
    logx=True,
    logy=True,
)
```

However, this does not defeat the purpose of the request I hope. I ran the following example similar to the SciPy example and the speedup is significant when using the size that `next_fast_len` provides for `numpy.fft.fft` (more than 6-fold). One can simply invoke SciPy when it is present in the Python environment, but it is not always available because it's a quite massive dependency.

```python
import numpy as np
from scipy.fft import fft as scipy_fft
from scipy.fft import next_fast_len

a = np.random.rand(93059)
optimum_fft_len = next_fast_len(len(a))

%timeit np.fft.fft(a)  # 4.75 ms ± 59.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit scipy_fft(a)  # 3.5 ms ± 83 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
%timeit np.fft.fft(a,n=optimum_fft_len)  # 741 µs ± 7.2 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
%timeit scipy_fft(a,n=optimum_fft_len)  # 996 µs ± 4.36 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by seberg** [comment]
I am curious if the overhead got worse in NumPy 2.1 (or 2.2/2.0, I don't recall?).

This seems like a simple enough helper in a small enough namespace, that I don't see much of a problem with adding it (even if of course this is really just duplicating exactly what scipy has).
OTOH, I am not sure how often it is actually used in practice.

**Comment by MothNik** [comment]
> This seems like a simple enough helper in a small enough namespace, that I don't see much of a problem with adding it (even if of course this is really just duplicating exactly what scipy has).
OTOH, I am not sure how often it is actually used in practice.

@seberg the duplication intensity of use are very good points!

Maybe, a hint to the SciPy function would already do the trick?

**Comment by seberg** [comment]
> Maybe a hint to the SciPy function in the documentation would already do the trick as a compromise?

Adding a doc reference to SciPy to "see also: and/or a `n` parameter seems like a clear improvement to me.  And if that seems good enough to you, then maybe that is all we should do.

**Comment by mhvk** [comment]
Note that `scipy` just exposes `good_size` from `pocketfft` - in principle, we could do the same.

It may also be an idea for `scipy` could just start using numpy's FFT implementation, since they depend on numpy anyway and we now use the same underlying code. Is it worth raising this possibility over at `scipy`?

**Comment by seberg** [comment]
Well, should maybe say again, that we can link the docs and I don't mind adding the function...

As to the observed speed difference: Should be careful to assume NumPy is faster, becuase I think that the overhead may have increased significantly with the move to gufuncs.
(SciPy has some additional layers that have the potential to make it much slower, but I don't think it would matter much unless you added some random imports of cupy or so.)

**Comment by MothNik** [comment]
@mhvk Regarding "escalating" this issue over to SciPy: The NumPy **v2.2** documentation states

<img width="736" alt="image" src="https://github.com/user-attachments/assets/96b4f7a8-007b-46c6-837b-5c6c20264a2d" />

So, is it really that simple to just make SciPy fully rely on NumPy? For me this seems like two different implementations under the hood.

@seberg I think adapting the docs would be a good compromise for now to avoid implementing something that then maybe needs to be adapted in the future again.

**Comment by rgommers** [comment]
I think that documentation needs a fix.

> So, is it really that simple to just make SciPy fully rely on NumPy? For me this seems like two different implementations under the hood.

I'll note that `numpy.fft` was only updated in numpy 2.0.0, and SciPy will need to support numpy 1.26 for ~1.5 years or so longer.

**Comment by mhvk** [comment]
Indeed, darn, that documentation is out of date. I should have updated it when I made the gufuncs for `float32` and `complex64`. @MothNik - since you found it, would you be willing to do the doc update? (if not, I'll do it, but perhaps this is a nice way to contribute!)

**Comment by MothNik** [comment]
@mhvk Thanks for the offer, but I'm really not that deep into what happens under the hood.
I assume that this paragraph can probably be removed completely, but tbh I don't feel comfortable editing a documentation where I don't truly know what's happening under the covers (especially here where an external low level library is involved).
